openmc/weight_windows.py

```diff
@@ -410,3 +411,182 @@ def from_hdf5(cls, group, meshes):
             weight_cutoff=weight_cutoff,
             id=id
         )
+
+def __wwinp_reader(path):
+    """
+    Generator that returns the next value in a wwinp file.
+
+    path : str or pathlib.Path
+        Location of the wwinp file
+    """
+
+    with open(path, 'r') as fh:
+
+        # read the first line of the file and
+        # keep only the first four entries
+        while True:
+            line = next(fh)
+            if line and not line.startswith('c'):
+                break
+
+        values = line.strip().split()[:4]
+        for value in values:
+            yield value
+
+        # the remainder of the file can be read as
+        # sequential values
+        while True:
+            line = next(fh)
+            # skip empty or commented lines
+            if not line or line.startswith('c'):
+                continue
+            values = line.strip().split()
+            for value in values:
+                yield value
```

Comment:
This kind of seems like overkill to me and results in lots of small I/O calls. I would recommend doing as much I/O as possible in one shot. Since the wwinp file is basically two lines of ints followed by a bunch of floats, something like:
```Python
with open(...) as wwinp:
    header = np.fromstring(wwinp.readline(), dtype=int, sep=' ')
    num_bins = np.fromstring(wwinp.readline(), dtype=int, sep=' ')
    data = np.fromstring(wwinp.read(), sep=' ')
```
Obviously you'll need a little more logic than that, but hopefully you get the idea. At that point, pulling information out is just a matter of selecting the right slice of those three arrays (rather than returning to the file to do more I/O through a call to a generator.

Reply:
One reason I took this approach was that it separates the I/O from the file specification logic. It seems that these would be more intermingled if we read them section-by-section or line-by-line. We'd need to read certain values of the header as `int`s and others as `float`s, for example, based on the mesh type. Maybe we would read all values with a floating point `dtype` and convert to `int` when appropriate?

It does keep that file handle open and makes a lot of I/O calls though, it's true. Would reading in all of the lines upfront make it a bit better?

Reply:
I do like how easy it would be to reshape the weight window values themselves when using `np.fromstring` though 🤔

Reply:
fwiw for the header line I would do something like this:
`with open(...) as wwinp:
    header = wwinp.readline().split(None, 4)
    _if, _iv, _ni, _nr = [int(x) for x in header[:4]]
    _probid = header[4] if len(header) > 4 else ""`

From there you know exactly how many lines to read based on those variables and where they should go using `np.fromstring`.

Reply:
:+1: to @eepeterson's suggestion -- it's a lot easier to follow and will be faster as well

Reply:
Alrighty I'll take it in that direction. Thanks @eepeterson!

Reply:
Okay just pushed the reader with the this new approach. It does look quite a bit better. Thanks for the help w this!